File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-1.java
class Solution {

    public int helper(int[] cost,int[] dp,int curr){
        

        if(dp[curr]!=0||curr<=1){
            return dp[curr];
        }

        int result=Math.min(helper(cost,dp,curr-1),helper(cost,dp,curr-2));
        dp[curr]=result+cost[curr];
        return dp[curr];
    }

    public int minCostClimbingStairs(int[] cost) {
        int n=cost.length;
        int[] dp=new int[n+1];
        dp[0]=cost[0];
        dp[1]=cost[1];
        if(n==2){
            return Math.min(dp[0],dp[1]);
        }
        return Math.min(helper(cost,dp,n-1),helper(cost,dp,n-2));

    }
}

File: Data Structures & Algorithms/house-robber/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        dp=[0]*len(nums)
        dp[0]=nums[0]
        if len(nums)>1:
            dp[1]=nums[1]
        else:
            return dp[0]
        
        def helper(curr):
            if curr<0:
                return 0
            if dp[curr]!=0:
                return dp[curr]
            res=max(helper(curr-2),helper(curr-3))
            dp[curr]=res+nums[curr]
            return dp[curr]
            
        return max(helper(len(nums)-1),helper(len(nums)-2))